packages/task/src/browser/quick-open-task.ts

```diff
@@ -187,36 +187,40 @@ export class QuickOpenTask implements QuickOpenModel, QuickOpenHandler {
         return `Task id: ${task.taskId}, label: ${task.config.label}`;
     }
 
-    private getFilteredTasks(recentTasks: TaskConfiguration[], configuredTasks: TaskConfiguration[], providedTasks: TaskConfiguration[])
-        : { filteredRecentTasks: TaskConfiguration[], filteredConfiguredTasks: TaskConfiguration[], filteredProvidedTasks: TaskConfiguration[] } {
+    private getFilteredTasks(recentTasks: TaskConfiguration[], configuredTasks: TaskConfiguration[], detectedTasks: TaskConfiguration[])
+        : { filteredRecentTasks: TaskConfiguration[], filteredConfiguredTasks: TaskConfiguration[], filteredDetectedTasks: TaskConfiguration[] } {
 
         const filteredRecentTasks: TaskConfiguration[] = [];
-        recentTasks.forEach(recent => {
-            const exist = [...configuredTasks, ...providedTasks].some(t => TaskConfiguration.equals(recent, t));
-            if (exist) {
-                filteredRecentTasks.push(recent);
+        for (const recent of recentTasks) {
+            const taskConfig = configuredTasks.find(task => recent.label === task.label) || detectedTasks.find(task => TaskConfiguration.equals(task, recent));
```

Comment:
like i mentioned in the other comment, using label comparison would cause issues that tasks from different sources are indentified as the same task.

This is what happened with your change:

![Peek 2019-06-27 09-38](https://user-images.githubusercontent.com/37082801/60270986-82ac0d80-98bf-11e9-8d04-6d316995f7a2.gif)

What i did was running `test task` from `test-source_A` and `test-source_B` respectively.
We ended up having only one `test task` in `recently used tasks`

============================

For your reference, this is what happened on master branch 

![Peek 2019-06-27 09-36](https://user-images.githubusercontent.com/37082801/60271140-cbfc5d00-98bf-11e9-91eb-5db0d3d54997.gif)

Reply:
We used to use `TaskConfiguration.equals()` which compares the `label` and `_source` - I guess you might have a good reason to use the label comparison instead. Could you please explain ?

Reply:
thank you, I'll check mentioned case

Reply:
I tried to fix the case like:
1. User runs detected task (the task moves to 'recently used' section)
2. User configures the task, edits it (adds some args for command, for example) and tries to run the task.
3. It's expected for me that modified task will be running, not detected task.
So at first I try to find configured task by label for ‘recently used’ section. Why by label? Because 
[according to the doc](https://github.com/theia-ide/theia/blob/deec2c4c04ec5b91c77bbad528810de5b6310ea2/packages/task/src/common/task-protocol.ts#L39-L44) for configured task source is the root folder, while for a detected task, it is the name of the provider - after configuring the task has another source - so I can not use equals method for this case. Is it correct? 
Then I use equals method to find task among detected tasks.

But you are right, for a multi-root use case it's not working, so we need to improve filtering tasks for 'recently used' section.

Reply:
I see. You are right: When the `detected task` is written into the `tasks.json`, it becomes a `configured task`. We need a machanism to differentiate the `configured detected task` from the `configured task`. And to be honest, I am not sure what is the best way to do.